src/box/txn_limbo.c

```diff
@@ -1236,10 +1316,8 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
 	case IPROTO_RAFT_PROMOTE:
 	case IPROTO_RAFT_DEMOTE: {
 		assert(limbo->svp_confirmed_lsn == -1);
-		limbo->svp_confirmed_lsn = limbo->confirmed_lsn;
-		limbo->confirmed_lsn = req->lsn;
-		vclock_reset(&limbo->confirmed_vclock, limbo->owner_id,
-			     req->lsn);
+		limbo->svp_confirmed_lsn = limbo->volatile_confirmed_lsn;
+		limbo->volatile_confirmed_lsn = req->lsn;
```

Comment:
You bump `volatile_confirmed_lsn`, which may be dangerous. I'm not sure, it's possible but let's anyway make sure, it cannot happen. Consider the following situation:

1. You failed to write `CONFIRM` in `txn_limbo_worker_f`. It'll cause `fiber_sleep(1)` and retry.
2. During `fiber_sleep` we try to write promote, which in `req_prepare` bumps `volatile_confirmed_lsn`. It's being written.
3. `txn_limbo_worker_f` wakes up and tries to write `CONFIRM` with lsn from the `promote`. There's no space in the journal queue, we're waiting.
4. `PROMOTE` is rolled back, we failed to write it, now there's enough space in queue, `CONFIRM` goes there and successfully written. It's incorrect `CONFIRM`.

The question here, whether 4th point can happen. But anyway, this bumping of `volatile_confirmed_lsn`  in `req_prepare` looks like a slippery slope to me

Reply:
What you described is impossible. We will never try to write CONFIRM concurrently to a PROMOTE request:
https://github.com/CuriousGeorgiy/tarantool/blob/f02774fad25d7cb67a619141a6a8cafe78887785/src/box/txn_limbo.c#L70-L71

Reply:
Thank you for explanation 💪